packages/assets-controllers/src/AssetsContractController.ts

```diff
@@ -62,137 +67,257 @@ export const SINGLE_CALL_BALANCES_ADDRESS_BY_CHAINID: Record<Hex, string> = {
-   * @param config - Initial options used to configure this controller.
-   * @param state - Initial state to set on this controller.
    */
-  constructor(
-    {
-      chainId: initialChainId,
-      onPreferencesStateChange,
-      onNetworkDidChange,
-      getNetworkClientById,
-    }: {
-      chainId: Hex;
-      onPreferencesStateChange: (
-        listener: (preferencesState: PreferencesState) => void,
-      ) => void;
-      onNetworkDidChange: (
-        listener: (networkState: NetworkState) => void,
-      ) => void;
-      getNetworkClientById: NetworkController['getNetworkClientById'];
-    },
-    config?: Partial<AssetsContractConfig>,
-    state?: Partial<BaseState>,
-  ) {
-    super(config, state);
-    this.defaultConfig = {
-      provider: undefined,
-      ipfsGateway: IPFS_DEFAULT_GATEWAY_URL,
-      chainId: initialChainId,
-    };
-    this.initialize();
-    this.getNetworkClientById = getNetworkClientById;
-
-    onPreferencesStateChange(({ ipfsGateway }) => {
-      this.configure({ ipfsGateway });
-    });
-
-    onNetworkDidChange(({ selectedNetworkClientId }) => {
-      const selectedNetworkClient = getNetworkClientById(
-        selectedNetworkClientId,
-      );
-      const { chainId } = selectedNetworkClient.configuration;
+  constructor({
+    messenger,
+    chainId: initialChainId,
+  }: {
+    messenger: AssetsContractControllerMessenger;
+    chainId: Hex;
+  }) {
+    this.messagingSystem = messenger;
+    this.#provider = undefined;
+    this.#ipfsGateway = IPFS_DEFAULT_GATEWAY_URL;
+    this.#chainId = initialChainId;
+
+    this.#registerActionHandlers();
+    this.#registerEventSubscriptions();
+  }
 
-      if (this.config.chainId !== chainId) {
-        this.configure({
-          chainId: selectedNetworkClient.configuration.chainId,
-        });
-      }
-    });
+  // TODO: Expand into base-controller utility function that batch registers action handlers.
+  #registerActionHandlers() {
+    const nonMethodClassProperties = [
+      'constructor',
+      'messagingSystem',
+      'setProvider',
+      'provider',
+      'ipfsGateway',
+      'chainId',
+    ];
+
+    getKnownPropertyNames<keyof this>(Object.getPrototypeOf(this)).forEach(
+      (method) => {
+        if (
+          ((key: keyof this): key is AssetsContractControllerMethodName =>
+            !nonMethodClassProperties.find((e) => e === key) &&
+            typeof this[key] === 'function')(method)
+        ) {
+          this.messagingSystem.registerActionHandler(
+            `${name}:${method}`,
+            // TODO: Write a generic for-loop implementation that iterates over an input union type in tandem with the input array.
+            // @ts-expect-error Both assigned argument and assignee parameter are using the entire union type for `method` instead of the type for the current element
+            this[method].bind(this),
+          );
+        }
+      },
+    );
+  }
```

Comment:
Proof of concept for new base-controller utility function that can batch register action handlers to a given messenger.